### ISSUE TITLE
crash

### ISSUE THREAD

**Issue by HenkieHanker2** [ar1a/Onion#6]
i get this and then it crashes, i tried debug but i get engine.pdb not loaded -,- 
![image](https://user-images.githubusercontent.com/35076607/38500385-8da9f232-3c0a-11e8-99aa-305e8e46c026.png)

**Comment by ar1a** [comment]
The offset must have changed.

**Comment by HenkieHanker2** [comment]
isn't it because `#pragma once
#include "BaseInterface.h"


class CClientMode extends BaseInterface
{

};


extern CClientMode* g_pClientMode;
` is empty?

**Comment by ar1a** [comment]
no, it's because https://github.com/ar1a/Onion/blob/master/Onion/Onion/InterfaceManager.cpp#L29 fails, which is because the offset changed and it can't dereference properly.

**Comment by HenkieHanker2** [comment]
would it work fine if i go use the sigscan for it?

**Comment by HenkieHanker2** [comment]
nvm

**Comment by HenkieHanker2** [comment]
i did it like this now `#include "InterfaceManager.h"
#include <iomanip>
#include <sstream>
#include "Log.h"
#include <Windows.h>
#include "Defines.h"
#include "interfaces\InterfaceClient.h"
#include "interfaces\InterfaceEngine.h"
#include "interfaces\InterfaceEntityList.h"
#include "interfaces\InterfaceTrace.h"
#include "interfaces\InterfaceClientMode.h"
#include "interfaces\InterfaceModelInfo.h"
#include "interfaces\InterfaceGlobals.h"
#include "Address.h"
#include "Sigscan.h"


CreateInterfaceFn get_module_factory1(HMODULE module)
{
	return reinterpret_cast<CreateInterfaceFn>(GetProcAddress(module, "CreateInterface"));
}

template<typename T>
T* get_interface1(CreateInterfaceFn f, const char* szInterfaceVersion)
{
	auto result = reinterpret_cast<T*>(f(szInterfaceVersion, nullptr));

	if (!result) {
		throw std::runtime_error(std::string("[get_interface] Failed to GetOffset interface: ") + szInterfaceVersion);
	}

	return result;
}

//CEngine*       g_pEngine = nullptr;
//CClient*	   g_pClient = nullptr;
//CClientMode*   g_pClientMode = nullptr;
//CEntityList*   g_pEntityList = nullptr;
//CTrace*        g_pTrace = nullptr;
//CModelInfo*    g_pModelInfo = nullptr;
//CGlobals*	   g_pGlobals = nullptr;


#include "Utils.h"

void InterfaceManager::GetInterfaces()
{

	auto engineFactory = get_module_factory1(GetModuleHandleW(L"engine.dll"));
	auto clientFactory = get_module_factory1(GetModuleHandleW(L"client.dll"));
	auto valveStdFactory = get_module_factory1(GetModuleHandleW(L"vstdlib.dll"));
	auto vguiFactory = get_module_factory1(GetModuleHandleW(L"vguimatsurface.dll"));
	auto vgui2Factory = get_module_factory1(GetModuleHandleW(L"vgui2.dll"));
	auto matSysFactory = get_module_factory1(GetModuleHandleW(L"materialsystem.dll"));
	auto dataCacheFactory = get_module_factory1(GetModuleHandleW(L"datacache.dll"));
	auto vphysicsFactory = get_module_factory1(GetModuleHandleW(L"vphysics.dll"));


	g_pClient = get_interface1<CClient>(clientFactory, "VClient018");
	g_pEngine = get_interface1<CEngine>(engineFactory, "VEngineClient014");
	g_pEntityList = get_interface1<CEntityList>(clientFactory, "VClientEntityList003");
	g_pTrace = get_interface1<CTrace>(engineFactory, "EngineTraceClient004");
	g_pModelInfo = get_interface1<CModelInfo>(engineFactory, "VModelInfoClient004");

	
	g_pClientMode = *(CClientMode**)(InterfaceManager::PatternScan(clientFactory, "A1 ? ? ? ? 8B 80 ? ? ? ? 5D") + 1);
	g_pGlobals = **(CGlobals***)(InterfaceManager::PatternScan(clientFactory, "A1 ? ? ? ? 5E 8B 40 10") + 1);


}


std::uint8_t* InterfaceManager::PatternScan(void* module, const char* signature)
{
	static auto pattern_to_byte = [](const char* pattern) {
		auto bytes = std::vector<int>{};
		auto start = const_cast<char*>(pattern);
		auto end = const_cast<char*>(pattern) + strlen(pattern);

		for (auto current = start; current < end; ++current) {
			if (*current == '?') {
				++current;
				if (*current == '?')
					++current;
				bytes.push_back(-1);
			}
			else {
				bytes.push_back(strtoul(current, &current, 16));
			}
		}
		return bytes;
	};

	auto dosHeader = (PIMAGE_DOS_HEADER)module;
	auto ntHeaders = (PIMAGE_NT_HEADERS)((std::uint8_t*)module + dosHeader->e_lfanew);

	auto sizeOfImage = ntHeaders->OptionalHeader.SizeOfImage;
	auto patternBytes = pattern_to_byte(signature);
	auto scanBytes = reinterpret_cast<std::uint8_t*>(module);

	auto s = patternBytes.size();
	auto d = patternBytes.data();

	for (auto i = 0ul; i < sizeOfImage - s; ++i) {
		bool found = true;
		for (auto j = 0ul; j < s; ++j) {
			if (scanBytes[i + j] != d[j] && d[j] != -1) {
				found = false;
				break;
			}
		}
		if (found) {
			return &scanBytes[i];
		}
	}
	return nullptr;
}




template<typename fn> fn InterfaceManager::GetInterface(std::string modulename, std::string interfacename)
{
	CreateInterfaceFn CreateInterface;
	ASSERT(CreateInterface =(CreateInterfaceFn)GetProcAddress(GetModuleHandle(modulename.c_str()), "CreateInterface"));
	
	fn pInterface = 0;
	for (int i = 100; i > 0; i--)
	{
		std::stringstream intf;
		intf << interfacename << std::setfill('0') << std::setw(3) << i;
		pInterface = (fn)(CreateInterface(intf.str().c_str(), 0));
		if (pInterface) { 
			Log::Write(DEBUG,"%s found 0x%.8X", intf.str().c_str(), pInterface); 
			break;
		}
	}

	return pInterface;
}
` gonna test now

**Comment by HenkieHanker2** [comment]
nop
